# Statistics for all Ojibwe treebanks

In [1]:
from collections import Counter

import pandas as pd
from conllu import parse_incr

from grammar_modules.disambiguation import REPO_ROOT
from corpus_view import analyze_corpus, get_token_statistics

In [2]:
TREEBANK_ROOT = REPO_ROOT / "data" / "treebanks"

TREEBANK_PATHS = {
    "Aanjikiing": TREEBANK_ROOT / "2026_08_18_Aanjikiing.conllu",
    "Akawe Niwii-tibaajim": TREEBANK_ROOT / "2026_08_18_Akawe_Niwii_tibaajim.conllu",
    "Anooj inaajimod": TREEBANK_ROOT / "2026_08_18_Anooj_inaajimod.conllu",
    "Awesiinyensag": TREEBANK_ROOT / "2026_08_18_Awesiinyensag.conllu",
    "Chi-mewinzha": TREEBANK_ROOT / "2026_08_18_Chi_mewinzha.conllu",
    "Gii-nitaa-aadisooke": TREEBANK_ROOT / "2026_08_18_Gii_nitaa_aadisooke_full.conllu",
    "Living Our Language": TREEBANK_ROOT / "2026_08_18_Living_our_Language.conllu",
    "Mino-doodaading": TREEBANK_ROOT / "2026_08_18_Mino_doodaading.conllu",
    "Naadamaading": TREEBANK_ROOT / "2026_08_18_Naadamaading.conllu",
    #"Oshkaabewis": TREEBANK_ROOT / "2026_08_18_Oshkaabewis.conllu",
    "Plums or Nuts": TREEBANK_ROOT / "2026_08_18_Plums_or_nuts.conllu",
    "Portage Lake": TREEBANK_ROOT / "2026_08_18_Portage_Lake.conllu",
    "The Dogs' Children": TREEBANK_ROOT / "2026_08_18_The_dogs_children.conllu",
    "Wiijikiiwending": TREEBANK_ROOT / "2026_08_18_Wiijikiiwending.conllu",
}

TREEBANK_PATHS

{'Aanjikiing': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Aanjikiing.conllu'),
 'Akawe Niwii-tibaajim': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Akawe_Niwii_tibaajim.conllu'),
 'Anooj inaajimod': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Anooj_inaajimod.conllu'),
 'Awesiinyensag': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Awesiinyensag.conllu'),
 'Chi-mewinzha': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Chi_mewinzha.conllu'),
 'Gii-nitaa-aadisooke': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_Constraint_Grammar/data/treebanks/2026_08_18_Gii_nitaa_aadisooke_full.conllu'),
 'Living Our Language': PosixPath('/Users/matthias/labs/ELF-Lab/ELF-Lab Repos/Ojibwe_

In [3]:
corpora = {}
token_statistics = {}

for corpus_name, corpus_path in TREEBANK_PATHS.items():
    print(f"Analyzing {corpus_name}...")
    corpora[corpus_name] = analyze_corpus(corpus_path=corpus_path)
    token_statistics[corpus_name] = get_token_statistics(corpus_path)

print("Finished.")

Analyzing Aanjikiing...
Analyzing Akawe Niwii-tibaajim...
Analyzing Anooj inaajimod...
Analyzing Awesiinyensag...
Analyzing Chi-mewinzha...
Analyzing Gii-nitaa-aadisooke...
Analyzing Living Our Language...
Analyzing Mino-doodaading...
Analyzing Naadamaading...
Analyzing Plums or Nuts...
Analyzing Portage Lake...
Analyzing The Dogs' Children...
Analyzing Wiijikiiwending...
Finished.


In [4]:
BASIC_UPOS_COLUMNS = (
    "VERB",
    "NOUN",
    "PROPN",
    "PRON",
    "ADV",
    "NUM",
    "DET",
)

basic_rows = []

for corpus_name in TREEBANK_PATHS:
    corpus = corpora[corpus_name]
    token_stats = token_statistics[corpus_name]
    upos_counts = token_stats["upos_counts"]
    sentence_count = len(corpus["sentences"])
    word_count = token_stats["word_count"]

    row = {
        "corpus": corpus_name,
        "sentences": sentence_count,
        "tokens_including_punctuation": token_stats["token_count"],
        "words_excluding_punctuation": word_count,
        "punctuation": token_stats["punctuation_count"],
        "unique_word_forms": token_stats["unique_word_forms"],
        "unique_lemmas": token_stats["unique_lemmas"],
        "mean_words_per_sentence": (
            round(word_count / sentence_count, 2)
        ),
    }
    row.update({upos.lower(): upos_counts[upos] for upos in BASIC_UPOS_COLUMNS})
    basic_rows.append(row)

basic_stats_df = pd.DataFrame(basic_rows).set_index("corpus")
basic_stats_df

,sentences,tokens_including_punctuation,words_excluding_punctuation,punctuation,unique_word_forms,unique_lemmas,mean_words_per_sentence,verb,noun,propn,pron,adv,num,det
corpus,,,,,,,,,,,,,,
Aanjikiing,745,8324,7211,1113,1840,1008,9.68,1873,777,18,519,2398,23,502
Akawe Niwii-tibaajim,2020,17498,13835,3663,4848,2224,6.85,4309,1869,17,824,4072,124,563
Anooj inaajimod,650,6321,5134,1187,1933,1073,7.90,1518,663,10,230,1569,39,188
Awesiinyensag,322,2341,1595,746,776,541,4.95,591,301,1,111,311,4,36
Chi-mewinzha,634,4426,3508,918,1264,774,5.53,972,458,4,409,1030,20,204
Gii-nitaa-aadisooke,2023,12808,9608,3200,3553,2529,4.75,2081,1070,1,509,2129,49,391
Living Our Language,2541,22270,17457,4813,5008,2581,6.87,4261,1693,55,1809,5371,122,702
Mino-doodaading,323,2368,1666,702,891,609,5.16,622,307,3,77,358,20,24
Naadamaading,364,2408,1657,751,833,573,4.55,587,344,3,85,346,14,38


## Simple overt argument position

In [5]:
for corpus_name, corpus in corpora.items():
    SV_count = 0
    VS_count = 0
    OV_count = 0
    VO_count = 0
    arguments = corpus["arguments"]
    for argument in arguments:
        if argument["relation"] == "nsubj":
            if argument["argument_token"]["id"] < argument["head_token"]["id"]:
                SV_count += 1
    print(f"{corpus_name}: {SV_count}")

Aanjikiing: 175
Akawe Niwii-tibaajim: 347
Anooj inaajimod: 120
Awesiinyensag: 82
Chi-mewinzha: 123
Gii-nitaa-aadisooke: 203
Living Our Language: 453
Mino-doodaading: 44
Naadamaading: 81
Plums or Nuts: 93
Portage Lake: 208
The Dogs' Children: 113
Wiijikiiwending: 181


	overt_subjects	SV	VS	subject_straddle	subject_unknown	overt_objects	OV	VO	object_straddle	object_unknown
corpus										
Aanjikiing	403	169	227	7	0	413	82	324	7	0
Akawe Niwii-tibaajim	785	339	438	8	0	811	215	585	11	0
Anooj inaajimod	300	116	180	4	0	287	57	229	1	0
Awesiinyensag	175	82	93	0	0	107	25	82	0	0
Chi-mewinzha	232	122	109	1	0	237	104	130	3	0
Gii-nitaa-aadisooke	603	201	400	2	0	244	79	165	0	0
Living Our Language	994	447	541	6	0	825	265	557	3	0
Mino-doodaading	136	44	92	0	0	131	17	113	1	0
Naadamaading	158	80	77	1	0	134	17	116	1	0
Plums or Nuts	234	91	138	5	0	202	56	143	3	0
Portage Lake	530	201	322	7	0	418	143	274	1	0
The Dogs' Children	280	112	166	2	0	149	38	110	1	0
Wiijikiiwending	332	179	151	2	0	230	42	186	2	0


Aanjikiing: 700
Akawe Niwii-tibaajim: 1388
Anooj inaajimod: 478
Awesiinyensag: 328
Chi-mewinzha: 492
Gii-nitaa-aadisooke: 812
Living Our Language: 1809
Mino-doodaading: 176
Naadamaading: 322
Plums or Nuts: 372
Portage Lake: 832
The Dogs' Children: 452
Wiijikiiwending: 721


## Simple S/V/O head order

In [7]:
from collections import defaultdict

for corpus_name, corpus in corpora.items():

    VSO_count = 0
    VOS_count = 0
    SVO_count = 0
    OVS_count = 0
    SOV_count = 0
    OSV_count = 0

    examples = defaultdict(list)

    for predicate in corpus["predicates"]:
        order = predicate["head_order"]

        if order == "VSO":
            VSO_count += 1
        elif order == "VOS":
            VOS_count += 1
        elif order == "SVO":
            SVO_count += 1
        elif order == "OVS":
            OVS_count += 1
        elif order == "SOV":
            SOV_count += 1
        elif order == "OSV":
            OSV_count += 1
        else:
            continue

        if len(examples[order]) < 3:
            examples[order].append({
                "predicate": predicate["predicate_token"]["form"],
                "text": predicate["text"],
                "eng": predicate["eng"],
            })

    print(f"\n{corpus_name}")
    print(f"VSO: {VSO_count}")
    print(f"VOS: {VOS_count}")
    print(f"SVO: {SVO_count}")
    print(f"OVS: {OVS_count}")
    print(f"SOV: {SOV_count}")
    print(f"OSV: {OSV_count}")

    print("\nExamples:")
    for order in ["VSO", "VOS", "SVO", "OVS", "SOV", "OSV"]:
        print(f"\n{order}:")
        for example in examples[order]:
            print(f"  Predicate: {example['predicate']}")
            print(f"  {example['text']}")
            if example["eng"]:
                print(f"  ENG: {example['eng']}")


Aanjikiing
VSO: 2
VOS: 9
SVO: 24
OVS: 5
SOV: 4
OSV: 2

Examples:

VSO:
  Predicate: ge-izhi-gikendamowaad
  Mii-go ge-izhi-gikendamowaad ingiw bezindookig , dibishkoo-go gaawiin gegoo imaa biinjina ayaamagasinoon ani-gaagiigidoyan .
  ENG: It sounds like they are reading a script.
  Predicate: da-ni-wiidoopamaad
  Mii dash iwidi da-ni-wiidoopamaad a'aw Waasigwan inow odinawemaaganan iwidi eyaanijin .
  ENG: And Waasigwan will share in a meal with all of his relatives that are over there.

VOS:
  Predicate: da-naadamaagod
  Mii-go gegapii a'aw mindimooyenyiban gii-ni-ozhibii'ang i'iw gaagiigidowin da-naadamaagod a'aw akiwenziiyiban gegoo aapiji da-ni-waniikesig imaa gaa-achigaadeg i'iw gaagiigidowining .
  ENG: Eventually that old lady wrote down the talk to help the old man not to forget what he was to say.
  Predicate: odaa-aanjitoosiin
  Gaawiin odaa-aanjitoosiin i'iw akeyaa gaa-izhi-gikinoo'amawind da-maajaa'iwed .
  ENG: ” Even though I know how to do these funerals I would not te

	verb_predicates	predicates_with_both_overt	SVO	SOV	VSO	VOS	OSV	OVS	other_both_overt
corpus									
Aanjikiing	5619	36	13	3	1	8	3	4	4
Akawe Niwii-tibaajim	30169	77	30	6	10	9	6	9	7
Anooj inaajimod	4554	27	7	1	4	6	3	2	4
Awesiinyensag	1773	20	10	0	2	3	1	4	0
Chi-mewinzha	6804	19	4	3	1	1	4	4	2
Gii-nitaa-aadisooke	6243	36	8	6	1	10	3	4	4
Living Our Language	29827	111	36	11	14	16	7	16	11
Mino-doodaading	1866	15	4	1	3	4	1	1	1
Naadamaading	1761	14	10	0	0	2	0	1	1
Plums or Nuts	4080	18	2	1	0	6	3	3	3
Portage Lake	16356	24	4	5	4	3	1	5	2
The Dogs' Children	3861	18	5	0	4	3	2	2	2
Wiijikiiwending	3324	52	29	2	5	7	2	5	2

## Inverse Queries

In [24]:
from collections import Counter

PROX = {"3SgProx", "3PlProx", "3Sg", "3Pl"}
OBV = {"3SgObv", "3PlObv"}

def has_relation(predicate, relation):
    return any(
        argument["relation"] == relation
        for argument in predicate["arguments"]
    )

def direction(predicate):
    p = predicate["predicate"]

    if p["subject_agreement"] in PROX and p["object_agreement"] in OBV:
        return "direct"

    if p["subject_agreement"] in OBV and p["object_agreement"] in PROX:
        return "inverse"

    return None


# Keep only non-local VTA predicates: proximate ↔ obviative
nonlocal_vtas = [
    predicate
    for predicate in all_predicates
    if predicate["predicate"]["verb_paradigm"] == "VTA"
    and (
        (
            predicate["predicate"]["subject_agreement"] in PROX
            and predicate["predicate"]["object_agreement"] in OBV
        )
        or
        (
            predicate["predicate"]["subject_agreement"] in OBV
            and predicate["predicate"]["object_agreement"] in PROX
        )
    )
]

no_overt = [
    p for p in nonlocal_vtas
    if not has_relation(p, "nsubj")
    and not has_relation(p, "obj")
]

subject_only = [
    p for p in nonlocal_vtas
    if has_relation(p, "nsubj")
    and not has_relation(p, "obj")
]

object_only = [
    p for p in nonlocal_vtas
    if not has_relation(p, "nsubj")
    and has_relation(p, "obj")
]

both_overt_nonlocal = [
    p for p in nonlocal_vtas
    if has_relation(p, "nsubj")
    and has_relation(p, "obj")
]

direction_counts = Counter(
    direction(p)
    for p in both_overt_nonlocal
)

print(f"Total non-local VTAs: {len(nonlocal_vtas)}")
print(f"No overt arguments: {len(no_overt)}")
print(f"Only overt subject: {len(subject_only)}")
print(f"Only overt object: {len(object_only)}")
print(f"Both arguments overt: {len(both_overt_nonlocal)}")

print("\nDirection counts when both overt")
print(f"Direct: {direction_counts['direct']}")
print(f"Inverse: {direction_counts['inverse']}")

Total non-local VTAs: 2276
No overt arguments: 904
Only overt subject: 292
Only overt object: 831
Both arguments overt: 249

Direction counts when both overt
Direct: 232
Inverse: 17


In [25]:
orders = ("SVO", "SOV", "VSO", "VOS", "OSV", "OVS")

direct = [
    predicate
    for predicate in both_overt_nonlocal
    if direction(predicate) == "direct"
]

inverse = [
    predicate
    for predicate in both_overt_nonlocal
    if direction(predicate) == "inverse"
]

direct_orders = Counter(
    predicate["head_order"]
    for predicate in direct
)

inverse_orders = Counter(
    predicate["head_order"]
    for predicate in inverse
)

for order in orders:
    print(
        order,
        "direct =", direct_orders[order],
        "inverse =", inverse_orders[order],
    )

SVO direct = 109 inverse = 2
SOV direct = 10 inverse = 0
VSO direct = 21 inverse = 2
VOS direct = 49 inverse = 1
OSV direct = 9 inverse = 2
OVS direct = 17 inverse = 8


In [ ]:
# These essentially need to be fixed/removed. 
# Also a big caveat on ditransitives, as ditransitives with 
# no overt iobj are currently being counted
other_orders = (
    set(direct_orders)
    | set(inverse_orders)
) - set(orders)

for order in sorted(other_orders):
    print(
        order,
        "direct =", direct_orders[order],
        "inverse =", inverse_orders[order],
    )

IOSVII direct = 1 inverse = 0
ISOV direct = 1 inverse = 0
OSVO direct = 2 inverse = 0
SOOV direct = 1 inverse = 0
SOVO direct = 1 inverse = 0
SSVO direct = 9 inverse = 0
SVOS direct = 1 inverse = 0
VOSS direct = 1 inverse = 0
VSOO direct = 0 inverse = 2
